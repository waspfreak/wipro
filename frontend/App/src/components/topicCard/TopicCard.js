import React, { useState, useCallback } from 'react';
import styles from './styles';
import { View, Text, TouchableOpacity, Image } from 'react-native';
import PropTypes from 'prop-types';
import ThumbnailInitials from '../thumbnailInitials/ThumbnailInitials';
import { iconPing } from '../../assets/images/index';
import { trimStringTolength } from '../../utils/stringUtils';

const useComponentSize = () => {
  const [state, setState] = useState({
    titleNumberOfLines: null,
    titleHeight: null
  });

  const onLayout = useCallback(event => {
    const numberOfLines = event.nativeEvent.lines.length;
    const totalHeight = event.nativeEvent.lines.reduce(
      (acc, line) => acc + line.height,
      0
    );
    setState({
      titleNumberOfLines: numberOfLines,
      titleHeight: totalHeight
    });
  }, []);

  return [state, onLayout];
};

const TopicCard = ({ title, description, onPress, members }) => {
  const totalMembers = members ? members.length : 1;
  const maxThumbnails = 4;
  const [state, onLayout] = useComponentSize();
  /* Title take precedence, the longer it is, the less description to be displayed */
  const showDescription = state.titleHeight < 50;
  const titleCharLimit = state.titleHeight > 50 ? 40 : 60;
  const summaryCharLimit = state.titleHeight > 50 ? 20 : 45;

  return (
    <TouchableOpacity style={styles.card} onPress={() => onPress(title)}>
      <View style={styles.topContainer}>
        <View style={styles.titleContainer}>
          <Text onTextLayout={onLayout} style={styles.title} testID="Title">
            {trimStringTolength(title, titleCharLimit)}
          </Text>
        </View>
        <TouchableOpacity>
          <View style={styles.iconContainer}>
            <Image source={iconPing} />
          </View>
        </TouchableOpacity>
      </View>
      {showDescription && (
        <View>
          <Text style={styles.summary}>
            {trimStringTolength(description, summaryCharLimit)}
          </Text>
        </View>
      )}
      <View style={styles.membersContainer}>
        <View style={styles.members}>
          {members &&
            members.map((member, index) => {
              return (
                index < maxThumbnails && (
                  <ThumbnailInitials key={index} name={member} size={25} />
                )
              );
            })}

          {members && totalMembers >= maxThumbnails && (
            <ThumbnailInitials
              key={totalMembers.length}
              name="test"
              size={25}
              extraMembers={totalMembers - maxThumbnails}
            />
          )}
        </View>
      </View>
    </TouchableOpacity>
  );
};

TopicCard.propTypes = {
  title: PropTypes.string.isRequired,
  description: PropTypes.string,
  onPress: PropTypes.func,
  members: PropTypes.array
};
export default TopicCard;
