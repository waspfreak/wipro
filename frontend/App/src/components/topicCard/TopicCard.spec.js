import React from 'react';
import TopicCard from './TopicCard';
import {
  render,
  NativeTestEvent,
  fireEvent,
  wait,
  waitForElement
} from '@testing-library/react-native';

describe('TopicCard tests', () => {
  let props;
  beforeEach(() => {
    props = {
      title: 'Title',
      description: 'this is a description',
      onPress: jest.fn(),
      members: ['first member', 'second test']
    };
  });

  it('should match TopicCard snapshot', () => {
    const { baseElement } = render(<TopicCard {...props} />);
    expect(baseElement).toMatchSnapshot();
  });

  // TODO: Needs improving this test
  it('should use layout event to work out height restrictions', async () => {
    const { baseElement, getByTestId } = render(<TopicCard {...props} />);
    const titleText = await waitForElement(() => getByTestId('Title'));
    fireEvent(
      baseElement,
      new NativeTestEvent('onLayout', {
        nativeEvent: { layout: { height: 100, lines: 3 } }
      })
    );
    await wait(() => expect(titleText).toBeTruthy());
  });
});
