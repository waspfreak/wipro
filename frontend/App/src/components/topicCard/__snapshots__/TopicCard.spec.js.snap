// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopicCard tests should match TopicCard snapshot 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <TouchableOpacity
      activeOpacity={0.2}
      style={
        Object {
          "backgroundColor": "#FFFFFF",
          "borderRadius": 8,
          "height": 140,
          "marginBottom": 10,
          "marginRight": 0,
          "marginTop": 10,
          "padding": 10,
          "paddingRight": 0,
          "shadowOffset": Object {
            "height": 1,
            "width": 0,
          },
          "shadowOpacity": 0.2,
          "shadowRadius": 1,
          "width": 168,
        }
      }
    >
      <View
        style={
          Object {
            "alignContent": "space-between",
            "flexDirection": "row",
          }
        }
      >
        <View
          style={
            Object {
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "flex-start",
              "maxHeight": 75,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#000000",
                "fontFamily": "SFProDisplay-Bold",
                "fontSize": 17,
                "marginBottom": 10,
                "maxWidth": 168,
              }
            }
            testID="Title"
          >
            Title
          </Text>
        </View>
        <TouchableOpacity
          activeOpacity={0.2}
        >
          <View
            style={
              Object {
                "flex": 1,
                "flexDirection": "row",
                "justifyContent": "flex-end",
              }
            }
          >
            <Image
              source={
                Object {
                  "testUri": "../../../App/src/assets/images/pin.png",
                }
              }
            />
          </View>
        </TouchableOpacity>
      </View>
      <View>
        <Text
          style={
            Object {
              "color": "#000000",
              "fontFamily": "SFProDisplay-Regular",
              "fontSize": 15,
              "paddingBottom": 10,
              "paddingRight": 10,
            }
          }
        >
          this is a description
        </Text>
      </View>
      <View
        style={
          Object {
            "flex": 1,
            "justifyContent": "flex-end",
          }
        }
      >
        <View
          style={
            Object {
              "flexDirection": "row",
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "borderRadius": 999999,
                  "justifyContent": "center",
                  "marginRight": 2,
                },
                Object {
                  "backgroundColor": "#6d6d6d",
                },
                Object {
                  "height": 25,
                  "width": 25,
                },
                null,
                null,
                null,
                undefined,
              ]
            }
            testID="userAvatarInitials"
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#FFFFFF",
                  },
                  Object {
                    "fontSize": 12,
                    "textAlign": "center",
                  },
                  Object {
                    "fontSize": 12.5,
                  },
                ]
              }
            >
              FM
            </Text>
          </View>
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "borderRadius": 999999,
                  "justifyContent": "center",
                  "marginRight": 2,
                },
                Object {
                  "backgroundColor": "#877457",
                },
                Object {
                  "height": 25,
                  "width": 25,
                },
                null,
                null,
                null,
                undefined,
              ]
            }
            testID="userAvatarInitials"
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#FFFFFF",
                  },
                  Object {
                    "fontSize": 12,
                    "textAlign": "center",
                  },
                  Object {
                    "fontSize": 12.5,
                  },
                ]
              }
            >
              ST
            </Text>
          </View>
        </View>
      </View>
    </TouchableOpacity>
  </View>
</View>
`;
